### PR TITLE
Integration of the ML based online Ecal DQM using autoencoder network [CMSSW_12_0_X]

### PR DESCRIPTION
**Pull request by abhih1**
#### PR description:

This PR introduces a machine learning based anomaly detection feature to the online ECAL DQM for the Ecal Barrel through a new DQM Client `MLClient`. An Autoencoder (AE) model with a ResNet architecture is trained on certified good data (EB digi occupancy) from 2018 runs.
On giving an input occupancy map to the AE, the encoder part of the AE encodes and learns the features and the decoder reconstructs the data from the encoded latent space to match the input as closely as possible. The reconstruction loss is then calculated, which is a mean squared error (MSE) between the input and output images at a tower level. Thus given an anomalous tower, the AE which has learnt the features of the good data will have a hard time reconstructing it and give a higher loss on the anomaly than on the good towers. A quality threshold is then applied on this loss map which marks it as Good or Bad, which is then stored as an ML quality summary plot.

We also do some pre processing and post processing steps to have a uniform, robust anomaly detection with fewer false positives over the changing detector conditions. The details of this are included in the comments throughout the code, and also can be found in this talk presented at a central DQM meeting [1]

Please note that the ML quality summary plot is a complimentary plot in addition to the existing DQM plots, and will be used for ECAL internal consumption only pending successful commissioning.
Along with the main ML quality summary plot, we also introduce a few monitoring plots to monitor the trends in Pileup (PU) per lumisection, no.of events processed in a lumisection and no.of events per input image to the ML model. This is in anticipation of the low luminosity expected during the start of Run 3. A few hyper parameters of our pre and post processing steps will also have to be tuned according to these trends, as currently they are derived from the conditions of Run 2.

 Please also note that this version of the PR is only meant for code integrations tests during the November 2021 MWGR. The plots will only be meaningful when we have real collisions, and modifications are foreseen after gathering data from the beginning of Run 3.
 
[1] https://indico.cern.ch/event/1042141/contributions/4377869/attachments/2254007/3824216/ML4DQM_AbhiramiHarilal_May28.pdf

#### PR validation:

The code was validated by running the modified online Ecal DQM configuration with the new client enabled and the resultant plots were examined by uploading the output file to a DQM test gui.
The new plots are confirmed and look reasonable.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
This is a backport of https://github.com/cms-sw/cmssw/pull/35990.
This is made to ensure that these changes are available for data taking during the November 2021 MWGR, which uses 12_0_X for production.

### CODE REVIEW

File: DQM/EcalMonitorClient/python/MLClient_cfi.py [hunk 1]
```diff
@@ -0,0 +1,411 @@
+import FWCore.ParameterSet.Config as cms
+
+from DQM.EcalMonitorTasks.OccupancyTask_cfi import ecalOccupancyTask
+
+#parameters dervied from training
+MLThreshold = 0.1761
+PUcorr_slope = 8625.62354249
+PUcorr_intercept= 354776.97564344
+#Avg Occupancy of 2018 data
+avgOcc = (1.4228867,1.4111477,1.4235557,1.403253 ,1.2926203,1.3110067,1.4317631,
```
Comment (jfernan2) [inline DQM/EcalMonitorClient/python/MLClient_cfi.py:10]: I am not sure if this is the preferred way to store this large information in CMSSW; I let @qliphy and @perrotta to comment

**Comment by jfernan2** [comment]
For the onnx file, I am afraid that normally for data files we have a dedicated package in https://github.com/cms-data corresponding to the cmssw package. For DQM there is for instance https://github.com/cms-data/DQM-SiStripMonitorClient and https://github.com/cms-data/DQM-DTMonitorClient . So i suggest to have a cms-data DQM-EcalMonitorClient package where the data files are moved. @smuzaffar What else should we add to get it working? A corresponding entry in CMSDIST, as https://github.com/cms-sw/cmsdist/blob/IB/CMSSW_10_2_X/gcc630/data-DQM-SiStripMonitorClient.spec right?

**Comment by smuzaffar** [comment]
@jfernan2 , just a data repo with a PR enough to test ( bot takes care of adding the new data package in cmsdist).

**Comment by jfernan2** [comment]
Thanks @smuzaffar 
Do we have rights to create such repo or do you take care of that?
Thank you

**Comment by smuzaffar** [comment]
I can  take care of it, are you happy with the new package name a `DQM/EcalMonitorClient` ? If yes then I can create the data repo

**Comment by jfernan2** [comment]
I think so, it goes in consonance with previous DQM package names, since at the end the Client is making use of this onnx file.
Please confirm us if/when @abhih1 should remove the onnx file from this PR and its forwardports to master and 12_1_X
Thank you very much

**Comment by smuzaffar** [comment]
https://github.com/cms-data/DQM-EcalMonitorClient is now available, please open a PR there for the new data file. cleanup this PR and remove the data file from the history

**Comment by smuzaffar** [comment]
@abhih1 , squash your two commits (to get rid of data file from git history) and force push your changes

**Comment by abhih1** [comment]
> @abhih1 , squash your two commits (to get rid of data file from git history) and force push your changes

@smuzaffar Thanks. I have done this now, and made a PR to https://github.com/cms-data/DQM-EcalMonitorClient.
 Just to understand, so the path to the onnx file mentioned in the client `DQM/EcalMonitorClient/onnxModels/resnet.onnx` should be changed or will it point to the data repository?

**Comment by smuzaffar** [comment]
@abhih1 , yes path should be `DQM/EcalMonitorClient/data/onnxModels/resnet.onnx`

**Comment by abhih1** [comment]
Thanks. I'll update this as well.

**Comment by jfernan2** [comment]
@smuzaffar it is not clear to me in which point we can trigger the tests, I do not see yet the cmsdist file created by the bot. Please let us know. Thanks

**Comment by smuzaffar** [comment]
please test with cms-data/DQM-EcalMonitorClient#1

**Comment by smuzaffar** [comment]
> @smuzaffar it is not clear to me in which point we can trigger the tests, I do not see yet the cmsdist file created by the bot. Please let us know. Thanks

bot automatically creates the cmsdist files, so once data pr is ready then one can already start testing

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
@abhih1 I see after the sucessful test that some plots are being added to Offline DQM too: is this intended? I thought you were looking for anomaly detection at Online DQM only
Thanks

**Comment by abhih1** [comment]
@jfernan2 yes we only want to do anomaly detection for online DQM but the Ecal DQM code is shared between online and offline DQM.
I have turned on the new client and added a switch for certain plot filling only for online configuration.

**Comment by abhih1** [comment]
It has already been included. 
The ML Client is switched on only for the online config here: https://github.com/cms-sw/cmssw/pull/35991/files#diff-7bb397ff9b2e1c01bdf725c05cf969f5fea5be030837d8417e90050f62f2f978R120

And the parameter for filling the PU plots here: https://github.com/cms-sw/cmssw/pull/35991/files#diff-7bb397ff9b2e1c01bdf725c05cf969f5fea5be030837d8417e90050f62f2f978R164

But there are some plots which will appear anyway in offline DQM, even though they wont be filled. I'm afraid we cant do anything about it. It will be buried in the Ecal workspace.

**Comment by jfernan2** [comment]
@smuzaffar since now the bot creates the changes in cmsdist in automatized way, we are not sure how to include this change in order to test this PR in Online DQM at P5

In the past we used to get the cmsdist PR and add it by hand, but now waiting for a new release implies to have this PR approved in advance, I am a bit confused.

Any advice is appreciated

**Comment by smuzaffar** [comment]
@jfernan2 , how do you normally test cmsdist changes in Online DQM at P5?

**Comment by jfernan2** [comment]
We used to get the cmsdist PR and check it at P5 for a manual test, last time was in 2017 though...
https://github.com/cms-sw/cmssw/pull/22970
with
https://github.com/cms-sw/cmsdist/pull/3980

But, for this particular case, since the module seems to be working well in Offline, perhaps the cmsdist can be merged and we wait for the next 12_0_X release to be installed at P5

**Comment by smuzaffar** [comment]
test parameters:
- pull_request = cms-sw/cmsdist#7433

**Comment by abhih1** [comment]
Hi @jfernan2  seems I was misinformed. Even if the client isn't disabled, those empty plots can be explicitly deleted from offline DQM. I'll push a commit for this soon.

**Comment by jfernan2** [comment]
Thanks @smuzaffar 
Despite the cmsdist PR is "linked" to master 12_2_X, can we add it on top of 12_0_3_patch1 that we are running at P5?

**Comment by jfernan2** [comment]
> Hi @jfernan2 seems I was misinformed. Even if the client isn't disabled, those empty plots can be explicitly deleted from offline DQM. I'll push a commit for this soon.

Thanks @abhih1 
While doing a preliminary check in Online DQM, we notice the plot in Ecal/Trends dubbed "Number of Events used per ML Image" is always empty, as in the offline plots. Is this expected?

**Comment by abhih1** [comment]
> Thanks @abhih1 While doing a preliminary check in Online DQM, we notice the plot in Ecal/Trends dubbed "Number of Events used per ML Image" is always empty, as in the offline plots. Is this expected?

Hi @jfernan2, That plot will only be filled after atleast 3 lumisections and 200 events of processing. So maybe during your testing this no.of lumisection condition is not met?

**Comment by jfernan2** [comment]
101LS and 1M events in playback:

https://cmsweb.cern.ch/dqm/online-playback/start?runnr=0;dataset=/Global/Online/ALL;sampletype=live;filter=all;referencepos=overlay;referenceshow=customise;referencenorm=True;referenceobj1=refobj;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Ecal;size=M;root=Ecal/Trends;focus=Ecal/Trends/Number%20of%20Events%20used%20per%20ML%20image;zoom=yes;

but plot empty

**Comment by abhih1** [comment]
Maybe because PU is -1 and this condition is not satisfied? : https://github.com/cms-sw/cmssw/pull/35991/files#diff-f6322d83172da74a703f6d7f55b6c0ee7950037ff3c3d56d89e2196d317d3595R39-R41

https://cmsweb.cern.ch/dqm/online-playback/start?runnr=0;dataset=/Global/Online/ALL;sampletype=live;filter=all;referencepos=overlay;referenceshow=customise;referencenorm=True;referenceobj1=refobj;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Ecal;size=M;root=Ecal/EventInfo;focus=Ecal/EventInfo/PU%20in%20the%20lumi;zoom=yes;

**Comment by jfernan2** [comment]
OK, I see, in that case it will be difficult to test that plot until real collisions arrive....

**Comment by abhih1** [comment]
yes unfortunately

**Comment by jfernan2** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
@jfernan2 , please use https://github.com/cms-sw/cmsdist/pull/7434 for 12.0.X testing

**Comment by jfernan2** [comment]
+1
Tested in Online DQM at P5

**Comment by qliphy** [comment]
+1

**Pull request merged**